### PR TITLE
[SystemZ][z/OS] __ptr32 support for z/OS

### PR DESCRIPTION
**Pull request by abhina-sree**
Enabling __ptr32 keyword to support in Clang for z/OS. It is represented by addrspace(1) in LLVM IR. Unlike existing implementation, __ptr32 is not mangled into symbol names for z/OS.

### CODE REVIEW

File: clang/lib/Basic/IdentifierTable.cpp [hunk 1]
```diff
@@ -211,6 +215,9 @@ static KeywordStatus getKeywordStatusHelper(const LangOptions &LangOpts,
   case KEYNOMS18:
     // The disable behavior for this is handled in getKeywordStatus.
     return KS_Unknown;
+  case KEYNOZOS:
+    // The disable behavior for this is handled in getKeywordStatus.
+    return KS_Unknown;
```
[the remark this answers is not in the document]
Comment (abhina-sree) [reply clang/lib/Basic/IdentifierTable.cpp]: Thanks, this is done!

File: clang/lib/Frontend/CompilerInvocation.cpp [hunk 2]
```diff
@@ -3650,6 +3650,14 @@ void CompilerInvocationBase::GenerateLangArgs(const LangOptions &Opts,
       GenerateArg(Consumer, OPT_ftrigraphs);
   }
 
+  if (T.isOSzOS()) {
+    if (!Opts.ZOSExt)
+      GenerateArg(Consumer, OPT_fno_zos_extensions);
+  } else {
+    if (Opts.ZOSExt)
+      GenerateArg(Consumer, OPT_fzos_extensions);
+  }
```
Comment (zibi2) [inline clang/lib/Frontend/CompilerInvocation.cpp:3659]: I think this can be simplified since zos-extensions are on by default on z/OS.

**Review by zibi2 — APPROVED** [review]
LGTM

File: clang/lib/Frontend/CompilerInvocation.cpp (hunk 2 above)
Comment (abhina-sree) [reply clang/lib/Frontend/CompilerInvocation.cpp]: Thanks, I refactored the code a bit

**Review by zibi2 — APPROVED** [review]
LGTM

**Comment by abhina-sree** [comment]
Enabling __ptr32 keyword to support in Clang for z/OS. It is represented by addrspace(1) in LLVM IR. Unlike existing implementation, __ptr32 is not mangled into symbol names for z/OS.

File: clang/lib/AST/ItaniumMangle.cpp [hunk 3]
```diff
@@ -2727,6 +2727,8 @@ void CXXNameMangler::mangleQualifiers(Qualifiers Quals, const DependentAddressSp
     //   <type> ::= U <OpenCL-addrspace>
     //   <type> ::= U <CUDA-addrspace>
 
+    llvm::Triple Triple = getASTContext().getTargetInfo().getTriple();
```
[the remark this answers is not in the document]
Comment (abhina-sree) [reply clang/lib/AST/ItaniumMangle.cpp:2730]: Thanks for catching, I fixed this here https://github.com/llvm/llvm-project/pull/102469

**Pull request merged**